OpenRA.Mods.Common/Traits/Buildings/Exit.cs

```diff
@@ -77,5 +77,29 @@ public static Exit RandomExitOrDefault(this Actor actor, World world, string pro
 			var shuffled = allOfType.Shuffle(world.SharedRandom);
 			return p != null ? shuffled.FirstOrDefault(p) : shuffled.First();
 		}
+	
+		public static Exit ClosestExitOrDefault(this Actor actor, World world, string productionType, Lazy<RallyPoint> rp, Func<Exit, bool> p = null)
+		{
+			if (!rp.Value.Sorted)
+			{
+				if (rp.Value.SortedExits == null)
+				{
+					rp.Value.SortedExits = Exits(actor, productionType);
+				}
+				if (!rp.Value.SortedExits.Any())
+					return null;
+				var sort = rp.Value.SortedExits.ToArray();
+				Array.Sort(sort, (x, y) => Distance(x.Info.ExitCell + actor.Location, rp.Value.Location).CompareTo(Distance(y.Info.ExitCell + actor.Location, rp.Value.Location)));
+				rp.Value.SortedExits = sort as IEnumerable<Exit>;
+				rp.Value.Sorted = true;
+			}
+
+			return (p != null ? rp.Value.SortedExits.FirstOrDefault(p) : rp.Value.SortedExits.First());
```

Comment:
This can all simplify down to something like
```csharp
public static Exit ClosestExitOrDefault(this Actor actor, World world, string productionType, CPos target, Func<Exit, bool> p = null)
{
	var allOfType = Exits(actor, productionType);
	if (!allOfType.Any())
		return null;

	return allOfType.MinBy(x => (x - target).LengthSquared);
}
```

Theres no need to tie this to `RallyPoint` specifically, so better to call this with the rallypoint location so that it can be reused by other code.